library/std/src/primitive_docs.rs

```diff
@@ -194,14 +193,48 @@ mod prim_bool {}
 /// # `!` and traits
 ///
 /// When writing your own traits, `!` should have an `impl` whenever there is an obvious `impl`
-/// which doesn't `panic!`. As it turns out, most traits can have an `impl` for `!`. Take [`Debug`]
+/// which doesn't `panic!`. The reason is that functions returning an `impl Trait` where `!`
+/// does not have an `impl` of `Trait` cannot diverge as their only possible code path. In other
+/// words, they can't return `!` from every code path. As an example, this code doesn't compile:
+///
+/// ```compile_fail
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     unimplemented!()
+/// }
+/// ```
+///
+/// But this code does:
+///
+/// ```
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     if true {
+///         unimplemented!()
+///     } else {
+///         0
+///     }
+/// }
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
+/// the `else` branch returns a `0`, which the compiler infers from the return type to be of type
+/// `u32`, which is a concrete type that `!` can be coerced to. See issue [#36375] for more
```

Comment:
```suggestion
/// `u32`. Since this is a concrete type, `!` gets coerced to `u32`. See issue [#36375] for more
```

Reply:
What about:

```suggestion
/// `u32`. Since `u32` is a concrete type, `!` can and will be coerced to it. See issue [#36375] for more
```

Reply:
Not a giant fan, but I don't think it's _less_ clear.

Reply:
"Since this is a concrete type" just sounded a bit confusing to me since it sounds kind of like we're referring to `!` and not `u32`.